ykrt/src/compile/j2/opt/cse.rs

```diff
@@ -0,0 +1,345 @@
+
+impl CSE {
+    /// Create an empty CSE object. As new instructions are committed to the eventual [Module],
+    /// they must also be `push`ed here so that CSE can be in-sync with the [Module].
+    pub(super) fn new() -> Self {
+        // Because we use `HEAD_UINT::MAX` as our "we haven't seen anything here yet" value, we
+        // have to make sure we won't overlap with a genuine value.
+        const {
+            assert!(InstIdx::MAX_INDEX > Inst::COUNT);
+        }
+        Self {
+            heads: [InstIdx::from_usize(InstIdx::MAX_INDEX); Inst::COUNT],
+            predecessors: Vec::new(),
+        }
+    }
+
+    pub(super) fn is_equiv(&self, opt: &dyn OptT, inst: &Inst) -> Option<InstIdx> {
+        // FIXME: This is a hack for "does this instruction have side effects".
+        if let Inst::Arg(_)
+        | Inst::Call(_)
+        | Inst::Const(_)
+        | Inst::Exit(_)
+        | Inst::Guard(_)
+        | Inst::Load(_)
+        | Inst::MemCpy(_)
+        | Inst::MemSet(_)
+        | Inst::Return(_)
+        | Inst::Store(_) = inst
+        {
+            return None;
+        }
+
+        #[cfg(test)]
+        {
+            if let Inst::BlackBox(_) = inst {
+                return None;
+            }
+        }
+
+        // FIXME: This is an inefficient way of working out what the maximum operand reference is,
+        // but `for_each_iidx` needs to be changed for that to be possible.
+        let max_ref = std::rc::Rc::new(std::cell::RefCell::new(InstIdx::from_usize(0)));
+        inst.for_each_iidx(|x| {
+            if x > *max_ref.borrow() {
+                *max_ref.borrow_mut() = x;
+            }
+        });
+        let max_ref = *max_ref.borrow();
+        let mut cur = self.heads[InstDiscriminants::from(inst) as usize];
+        while cur != InstIdx::MAX_INDEX && cur >= max_ref {
+            let equiv = opt.equiv_iidx(cur);
+            if equiv >= max_ref && opt.inst(equiv).cse_eq(opt, inst) {
+                return Some(equiv);
+            }
+            cur = self.predecessors[usize::from(cur)];
+        }
+        None
+    }
+
+    /// As a [Module] is being built-up, this method must be called with each new instruction
+    /// generated.
+    pub(super) fn push(&mut self, inst: &Inst) {
+        let dim_off = InstDiscriminants::from(inst) as usize;
+        let prev = self.heads[dim_off];
+        self.predecessors.push(prev);
+        self.heads[dim_off] = InstIdx::from_usize(self.predecessors.len() - 1);
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::compile::j2::opt::{
+        opt::{OptOutcome, test::opt_and_test},
+        strength_fold::strength_fold,
+    };
+    use std::{cell::RefCell, rc::Rc};
+
+    fn test_cse(mod_s: &str, ptn: &str) {
+        let cse = Rc::new(RefCell::new(CSE::new()));
+        opt_and_test(
+            mod_s,
+            |opt, mut inst| {
+                if let Inst::Guard(_) = inst {
+                    match strength_fold(opt, inst) {
```

Comment:
I'm a little torn whether we want strength reduction to be in the CSE test. That would make it possible for strength reduction to hide a CSE bug. Ideally, a unit test should just test the CSE itself?

We can have a larger integration test that tests the whole optimization suite later.

Reply:
A naiive question: why can't we optimise across that assuming same SSA variables and the operations are "pure". E.g.
```
          %0: i8 = arg [reg]
          %1: i8 = arg [reg]
          %4: i8 = add %0, %0
          %2: i1 = icmp eq %0, %1
          guard true, %2, []
          %5: i8 = add %1, %1
```
We should be able to CSE `%5` with `%4`?

Reply:
IIUC, that's what the test starting at line 169 shows.

Reply:
Ah ok, that makes more sense now! You need the equivalences to be built before the CSE pass IIUC?

Reply:
Ah I misunderstood the test then. Thank you. This makes sense. Basically the guard just says `%1 == %0` after the `icmp` and CSE can operate on that. Neat.